storage/pruning/pruningStorer.go

```diff
@@ -0,0 +1,444 @@
+				base64.StdEncoding.EncodeToString(key), ps.identifier)
+		}
+	}
+
+	return v.([]byte), nil
+}
+
+// GetFromEpoch will search a key only in the persister for the given epoch
+func (ps *PruningStorer) GetFromEpoch(key []byte, epoch uint32) ([]byte, error) {
+	// TODO: this will be used when requesting from resolvers
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	v, ok := ps.cacher.Get(key)
+	if ok {
+		return v.([]byte), nil
+	}
+
+	persisterData, exists := ps.persistersMapByEpoch[epoch]
+	if !exists {
+		return nil, fmt.Errorf("key %s not found in %s",
+			base64.StdEncoding.EncodeToString(key), ps.identifier)
+	}
+
+	if !persisterData.isClosed {
+		return persisterData.persister.Get(key)
+	} else {
+		persister, err := ps.persisterFactory.Create(persisterData.path)
+		if err != nil {
+			log.Debug("open old persister", "error", err.Error())
+			return nil, err
+		}
+
+		err = persister.Init()
+		if err != nil {
+			log.Debug("init old persister", "error", err.Error())
+			return nil, err
+		}
+
+		res, errToRet := persister.Get(key)
+
+		err = persister.Close()
+		if err != nil {
+			log.Debug("close old persister", "error", err.Error())
+			return nil, err
+		}
+
+		if errToRet == nil {
+			return res, nil
+		}
+		log.Warn("get from closed persister",
+			"id", ps.identifier,
+			"epoch", epoch,
+			"key", key,
+			"error", errToRet.Error())
+
+		return nil, fmt.Errorf("key %s not found in %s",
+			base64.StdEncoding.EncodeToString(key), ps.identifier)
+	}
+}
+
+// Has checks if the key is in the Unit.
+// It first checks the cache. If it is not found, it checks the bloom filter
+// and if present it checks the db
+func (ps *PruningStorer) Has(key []byte) error {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	has := ps.cacher.Has(key)
+	if has {
+		return nil
+	}
+
+	if ps.bloomFilter == nil || ps.bloomFilter.MayContain(key) {
+		for _, persister := range ps.activePersisters {
+			if persister.persister.Has(key) != nil {
+				continue
+			}
+
+			return nil
+		}
+	}
+
+	return storage.ErrKeyNotFound
+}
+
+// HasInEpoch checks if the key is in the Unit in a given epoch.
+// It first checks the cache. If it is not found, it checks the bloom filter
+// and if present it checks the db
+func (ps *PruningStorer) HasInEpoch(key []byte, epoch uint32) error {
+	// TODO: this will be used when requesting from resolvers
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	has := ps.cacher.Has(key)
+	if has {
+		return nil
+	}
+
+	if ps.bloomFilter == nil || ps.bloomFilter.MayContain(key) {
+		persister, ok := ps.persistersMapByEpoch[epoch]
+		if !ok {
+			return storage.ErrKeyNotFound
+		}
+
+		return persister.persister.Has(key)
+	}
+
+	return storage.ErrKeyNotFound
+}
+
+// Remove removes the data associated to the given key from both cache and persistence medium
+func (ps *PruningStorer) Remove(key []byte) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	ps.cacher.Remove(key)
+	return ps.activePersisters[0].persister.Remove(key)
```

Comment:
how do you know the key is in the first active persister?

it should be removed from any of the active persisters iterating from most recent to older and break on found.
I could think of the situation where you are right after the start of epoch event and you would need to rollback your chain to previous epoch blocks.

Reply:
Changed. Now I iterate over all the active persisters and if i get a nil error then it means that the key-value pair was removed. otherwise I try in the next persisters and if no one is successful then I return the last error